### PR TITLE
Refactor bump code

### PR DESCRIPTION
**Pull request by dappnodedev**
Refactor bump code to prepare the action to receive `--template` flag

### CODE REVIEW

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { Manifest, Compose } from "@dappnode/types";
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:1]: This file is confusing, has arbitrary functions in it, either find a better file struct or create the required modules to achieve a better organization

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 2]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
+    gitSettings,
+    composeVersionsToUpdate,
+    githubSettings,
+  });
 
-Manifest - ${JSON.stringify(manifest, null, 2)}
+  try {
+    await closeOldPrs(repo, branchName);
+  } catch (e) {
+    console.error("Error on closeOldPrs", e);
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:95]: Since you are not throwing I would use rather `console.warn`

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 3]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
 
-  const ethProviderAvailable = await getFirstAvailableEthProvider({
-    providers: ethProviders
-  });
+  return versionsToUpdate;
+}
 
-  if (!ethProviderAvailable)
-    throw Error(`No eth provider available. Tried: ${ethProviders.join(", ")}`);
+async function updateManifest({
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:163]: This function has a lot of logic embedded, could you add a detailed description of what it does, what it takes and what it returns

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 4]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
-        };
-        // Use a Map since there may be multiple matches for the same argName
-        versionsToUpdateMap.set(repoSlug, {
-          repoSlug,
-          newVersion,
-          currentVersion
-        });
-      }
+function updateComposeVersions(dir: string, compose: Compose, upstreamRepoVersions: UpstreamRepoMap): ComposeVersionsToUpdate {
+  const newCompose = JSON.parse(JSON.stringify(compose)); // Deep copy
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:134]: why is this needed?

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 5]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
+  const composeVersionsToUpdate = updateComposeVersions(dir, compose, upstreamRepoVersions);
+  if (isEmpty(composeVersionsToUpdate)) {
+    console.log("All versions are up-to-date");
+    return;
+  }
 
-  const ethProviders = use_fallback
-    ? [ethProvider, ...defaultEthProviders.filter(p => p !== ethProvider)]
-    : [ethProvider];
+  updateManifest({ manifest, manifestFormat: format, composeVersionsToUpdate, dir, ethProvider });
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:83]: missing `await`

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 6]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
-Manifest - ${JSON.stringify(manifest, null, 2)}
+  try {
+    await closeOldPrs(repo, branchName);
+  } catch (e) {
+    console.error("Error on closeOldPrs", e);
+  }
 
-Compose - ${JSON.stringify(compose, null, 2)}
-`);
+  const gitHead = await getGitHead();
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/index.ts:98]: if you are not using `gitHead` nowehere else inside this loop why not calling it directly from `buildAndComment`

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  
+    Upstream Settings - ${JSON.stringify(upstreamSettings, null, 2)}
+  
+    Git Settings - ${JSON.stringify(gitSettings, null, 2)}
+    
+    Manifest - ${JSON.stringify(manifest, null, 2)}
+    
+    Compose - ${JSON.stringify(compose, null, 2)}
+    
+    ETH Provider - ${ethProvider}
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:96]: would be interesting to print also the ipfs provider for debugging

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+export async function readInitialSetup({ dir, userEthProvider, useFallback }: { dir: string, userEthProvider: string, useFallback: boolean }): Promise<InitialSetupData> {
+    const envFileArgs = readBuildSdkEnvFileNotThrow(dir);
+
+    const { manifest, format } = readManifest({ dir });
+    const compose = readCompose({ dir });
+
+    const upstreamRepos = envFileArgs
+        ? [envFileArgs._BUILD_UPSTREAM_REPO]
+        : parseCsv(manifest.upstreamRepo);
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:17]: could you add a todo with a link reference to https://github.com/dappnode/DAppNodeSDK/issues/408 saying that the upstreamRepo format will change

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    const upstreamArgs = envFileArgs
+        ? [envFileArgs._BUILD_UPSTREAM_VERSION]
+        : parseCsv(manifest.upstreamArg || "UPSTREAM_VERSION");
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:20]: same as above

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import { getFirstAvailableEthProvider } from "../../../utils/tryEthProviders.js";
+
+export async function readInitialSetup({ dir, userEthProvider, useFallback }: { dir: string, userEthProvider: string, useFallback: boolean }): Promise<InitialSetupData> {
+    const envFileArgs = readBuildSdkEnvFileNotThrow(dir);
+
+    const { manifest, format } = readManifest({ dir });
+    const compose = readCompose({ dir });
+
+    const upstreamRepos = envFileArgs
+        ? [envFileArgs._BUILD_UPSTREAM_REPO]
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:16]: Ive never seen this env, can you either research if it can be removed or add a todo to do so

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        : parseCsv(manifest.upstreamRepo);
+    const upstreamArgs = envFileArgs
+        ? [envFileArgs._BUILD_UPSTREAM_VERSION]
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:19]: same as above

File: src/commands/githubActions/bumpUpstream/setup.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    const ethProviders = getEthProviders(useFallback, userEthProvider);
+
+    const ethProviderAvailable = await getFirstAvailableEthProvider({
+        providers: ethProviders
+    });
+
+    if (!ethProviderAvailable)
+        throw Error(`No eth provider available. Tried: ${ethProviders.join(", ")}`);
+
+    return {
```
Comment (pablomendezroyo) [inline src/commands/githubActions/bumpUpstream/setup.ts:40]: For me it looks too arbitrary this data return. I would rather try to organize it by topics, a module like:
```
| settings
| -- index.js
| -- ethProvider
|  -- | -- index.js
|  -- | -- getEthProviders.js
| -- dappnodeFiles
|  -- | -- index.js
|  -- | -- getManifest.js
|  -- | -- getCompose.js
| -- upstream
|  -- | -- index.js
|  -- | -- validateUpstreamData.js
|  -- | -- getUpstreamData.js
| -- git
|  -- | -- index.js
```

**Review by pablomendezroyo — CHANGES_REQUESTED** [review]

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 13]
```diff
@@ -1,33 +1,27 @@
...
+import { isEmpty } from "lodash-es";
+import { UpstreamSettings, UpstreamRepoMap, ComposeVersionsToUpdate, GitSettings, GithubSettings } from "./types.js";
+import { closeOldPrs, getGitHubSettings, getPrBody, getUpstreamVersionTag, isBranchNew, isValidRelease } from "./git.js";
+import { printSettings, readInitialSetup } from "./setup.js";
+import { ManifestFormat } from "../../../files/manifest/types.js";
 
 interface CliCommandOptions extends CliGlobalOptions {
   eth_provider: string;
   use_fallback: boolean;
+  template: boolean;
```
[the remark this answers is not in the document]
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:24]: This is for another PR, sorry. Already removed

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 6 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:98]: You are right, done

Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:98]: Because `buildAndComment` is also used in the build command and it is not always the git head

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 14]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
-Compose - ${JSON.stringify(compose, null, 2)}
-`);
+  const gitHead = await getGitHead();
+  await buildAndComment({ dir, commitSha: gitHead.commit, branch: branchName });
+}
 
-  if (upstreamRepos.length < 1) {
-    throw Error("Must provide at least one 'upstream_repo'");
-  }
+async function getUpstreamRepoVersions(upstreamSettings: UpstreamSettings[]): Promise<UpstreamRepoMap> {
```
[the remark this answers is not in the document]
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:102]: Okay I agree

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 4 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:134]: It is used to update the manifest and to open the PR

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 15]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
-          }
-        };
-        // Use a Map since there may be multiple matches for the same argName
-        versionsToUpdateMap.set(repoSlug, {
-          repoSlug,
-          newVersion,
-          currentVersion
-        });
-      }
+function updateComposeVersions(dir: string, compose: Compose, upstreamRepoVersions: UpstreamRepoMap): ComposeVersionsToUpdate {
```
[the remark this answers is not in the document]
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:133]: Done!

File: src/commands/githubActions/bumpUpstream/setup.ts (hunk 7 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/setup.ts:96]: It is not possible to pass a flag for the IPFS provider in this command. It is hardcoded in the `buildAndComment` function:

```
  const { releaseMultiHash } = await buildHandler({
    provider: "pinata",
    upload_to: "ipfs",
    require_git_data: true,
    delete_old_pins: true,
    verbose: true
  });
  ```

So it is not possible to print the IPFS provider here

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 3 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:163]: Okay!

File: src/commands/githubActions/bumpUpstream/index.ts [hunk 16]
```diff
@@ -55,180 +49,178 @@ export const gaBumpUpstream: CommandModule<
...
-  const templateArgs = readBuildSdkEnvFileNotThrow(dir);
 
-  const { manifest, format } = readManifest({ dir });
-  const compose = readCompose({ dir });
+  const { upstreamSettings, manifestData: { manifest, format }, compose, gitSettings, ethProvider } = await readInitialSetup({ dir, userEthProvider, useFallback });
+  printSettings(upstreamSettings, gitSettings, manifest, compose, ethProvider);
 
-  const upstreamRepos = templateArgs
-    ? [templateArgs._BUILD_UPSTREAM_REPO]
-    : parseCsv(manifest.upstreamRepo);
```
[the remark this answers is not in the document]
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/index.ts:74]: Why do you say this? It is being used

File: src/commands/githubActions/bumpUpstream/setup.ts (hunk 10 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/setup.ts]: Removed!

File: src/commands/githubActions/bumpUpstream/setup.ts (hunk 12 above)
Comment (dappnodedev) [reply src/commands/githubActions/bumpUpstream/setup.ts]: I did it in a simpler way, but I implemented it in module format

File: src/commands/githubActions/bumpUpstream/index.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (pablomendezroyo) [reply src/commands/githubActions/bumpUpstream/index.ts:74]: true my bad

**Review by pablomendezroyo — APPROVED** [review]
